arch/risc-v/src/common/riscv_cpuindex.c

```diff
@@ -48,10 +50,14 @@
  *
  ****************************************************************************/
 
-int up_cpu_index(void)
+uintptr_t riscv_cpuindex(void)
 {
-  int mhartid;
+  return READ_CSR(mhartid);
+}
 
-  asm volatile ("csrr %0, mhartid": "=r" (mhartid));
-  return mhartid;
+#ifdef CONFIG_SMP
+int up_cpu_index(void)
+{
+  return (int)riscv_cpuindex();
```

Comment:
This should just become a tail call and is not dependent on preprocessor substitution, would this solution suffice ?

Reply:
This can be a solution. Sorry I'm not fully aware about the values that `mhartid` can carry, so I'm just interested can we go with
```
int up_cpu_index(void)
{
  uintptr_t mhartid = READ_CSR(mhartid);

  return mhartid;
}
```
or even with
```
int up_cpu_index(void)
{
  return READ_CSR(mhartid);
}
```
?? do we need the upper bits? because if the values are `0..50` for example, then maybe we can reuse existing API and not introduce `riscv_cpuindex`.

Reply:
mhartid is XLEN wide. IMO there is no need to strip or splice it.

The reason for having a separate function for obtaining hartid is because this function will look like this eventually:

`uintptr_t riscv_cpuindex(void)
{
#ifdef CONFIG_ARCH_USE_S_MODE
  /* Kernel is in S-mode */

  return riscv_mcall_mhartid();

#else
  /* Kernel is in M-mode */

  return READ_CSR(mhartid);
#endif
}`